Add ADR about the application architecture (#210)
DFE-Digital/apply-for-postgraduate-teacher-training: adr/0004-application-architecture.md

# 4. Application architecture

Date: 2019-09-25

## Status

Accepted

## Context

> We are starting to build a beta of the 'Apply for postgraduate teacher training' service. This service will have 3 distinct user groups, who will have their own interface in the service.

Reviewer: What about referees?

> - Candidates are people who would like to become a teacher. The service

Reviewer: This sentence is incomplete

> - Candidates are people who would like to become a teacher. The service

Author: ```suggestion
- Candidates are people who would like to become a teacher.
```

- Provider users work for an organisation that provides teacher training, like a school or university
- Vendors make software that universities use for administration

For the architecture of the system, we've discussed 2 options: a microservice architecture and a monolith.

In the microservice architecture, we might split up the service into 3 or more applications - for example an API, a frontend for candidates and a frontend for the provider users. In the monolithic architecture, all components live together in a single application.

## Decision

> We've chosen to go for a monolith. After discussing the options, we hypothesise that the microservice architecture for this project doesn't provide many of the benefits often associated with microservice approach (for [example](https://rubygarage.org/blog/advantages-of-microservices)). For example, the components are very tightly coupled and will almost always change together rather than independently. Using a monolith approach will make the project easier to run for developers and easier to deploy.

Reviewer 2: `[example]` isn't great link text, maybe `[example advantages of microservices]`?

> We've chosen to go for a monolith. After discussing the options, we hypothesise that the microservice architecture for this project doesn't provide many of the benefits often associated with microservice approach (for [example](https://rubygarage.org/blog/advantages-of-microservices)). For example, the components are very tightly coupled and will almost always change together rather than independently. Using a monolith approach will make the project easier to run for developers and easier to deploy.

Author: ```suggestion
We've chosen to go for a monolith. After discussing the options, we hypothesise that the microservice architecture for this project doesn't provide many of the [benefits often associated with microservice approach](https://rubygarage.org/blog/advantages-of-microservices).

For example, the components are very tightly coupled and will almost always change together rather than independently. Using a monolith approach will make the project easier to run for developers and easier to deploy.
```

> We've chosen to go for a monolith. After discussing the options, we hypothesise that the microservice architecture for this project doesn't provide many of the benefits often associated with microservice approach (for [example](https://rubygarage.org/blog/advantages-of-microservices)). For example, the components are very tightly coupled and will almost always change together rather than independently. Using a monolith approach will make the project easier to run for developers and easier to deploy.

Reviewer: Should we acknowledge and dismiss the supposed benefits of microservice architecture, being (imo, and I'm a novice):

- independent teams can develop independent ways of looking at the same core domain (this domain is too small for that)
- as teams' (ie components') priorities start to diverge the communication overhead of a monolith eventually becomes unbearable (I think you're saying that the teams won't separate very much?)

However, there are problems with a monolith as well. In particular, it can lead to an application that is hard to understand because of a wild growth of classes and files. Additionally, the tests on a large monolith may become slow over time, causing frustration and slowdown in development.

## Consequences

> - We need to make sure we structure the appliction well

Author: ```suggestion
- We need to make sure we structure the application well
```

Reviewer: This is a bit vague. How about sth like "where boundaries exist between responsibilities we will try to maintain them by using modules and other units of separation"?

# 4. Application architecture

Date: 2019-09-25

## Status

Accepted

## Context

We are starting to build a beta of the 'Apply for postgraduate teacher training' service. This service will have 5 distinct user groups, who will have their own interface in the service.

1. Candidates are people who would like to become a teacher.
2. Provider users work for an organisation that provides teacher training, like a school or university
3. Vendors make software that universities use for administration

> 4. Referees will be asked by candidates to provide a reference

Reviewer 3: ```suggestion
4. Referees will be provided by candidates and the service will ask them to provide a reference
```

I might be wrong, but I believe this is more accurate. The `asked by candidates` suggests it's a thing that a candidate will have to do but from my understanding it's a thing we'll handle.

5. Users within DfE need information about the performance of the service and solve problems

For the architecture of the system, we've discussed 2 options: a microservice architecture and a monolith.

In the microservice architecture, we might split up the service into 3 or more applications - for example an API, a frontend for candidates and a frontend for the provider users. In the monolithic architecture, all components live together in a single application.

## Decision

We've chosen to go for a monolith. After discussing the options, we hypothesise that the microservice architecture for this project doesn't provide many of the [benefits often associated with microservice approach](https://rubygarage.org/blog/advantages-of-microservices).

For example:

> - Teams can work indepedently: we anticipate that the team size of this project will be limited (less than 20), so communication is less of a problem

Reviewer 3: ```suggestion
- Teams can work independently: we anticipate that the team size of this project will be limited (less than 20), so communication is less of a problem
```
